runtime/src/accounts_db.rs

```diff
@@ -650,7 +655,7 @@ impl AccountsDB {
     // collection
     // Only remove those accounts where the entire rooted history of the account
     // can be purged because there are no live append vecs in the ancestors
-    pub fn clean_accounts(&self) {
+    pub fn clean_accounts(&self, max_clean_root: Option<Slot>) {
```

Comment:
this might be personal taste, but my favorite way of changing widely-reference method's signature is like this to reduce diff:


```rust
  fn clean_accounts(&self) {
    self.do_clean_accounts(None);
  }
  fn clean_accounts_upto_exclusive(&self, max_...: Slot) {
    self.do_clean_accounts(Some(max_....saturating_sub(1)));
  }
  fn do_clean_accounts(&self, max...: Option<Slot>) {
     ....
  }
```

Reply:
I generally try to avoid the `do_*` where possible b/c I think it makes the code a little more bloated and adds a layer of indirection

Reply:
I see your point in `do_*`. So, let's take the tree-wide rewrite of `clean_accounts(None)`.

Lastly, I want to do something for the handling of `- 1`. I bet engineers always get confused a bit for the `- 1` at the callsite here and there without proper commenting. The confusion would be like `why this needs to be - 1, why not 0, -2?` and `why bank's slot needs to be passed redundantly? Is there valid usecase for passing something different?`:

```
$ git grep clean_accounts.*Some
ledger-tool/src/main.rs:                    bank.clean_accounts(Some(bank.slot() - 1));
runtime/src/accounts_background_service.rs:                snapshot_root_bank.clean_accounts(Some(snapshot_root_bank.slot() - 1));
runtime/src/bank.rs:        self.clean_accounts(Some(self.slot() - 1));
runtime/src/bank.rs:            self.clean_accounts(Some(self.slot() - 1));
runtime/src/bank.rs:        bank1.clean_accounts(Some(1));

```

I can observe that almost all callsites are just doing the common pattern of `self.clean_accounts(self.slot() - 1)`. So, how about abstracting away by the `- 1` by en-capsuling inside `Bank` at https://github.com/solana-labs/solana/pull/12194/files#r495566668?:

```rust
struct Bank {
    fn clean_accounts(&self, skip_last: bool) {
        let max_clean_slot = if full {
            None
        } else {
            // foo bar why this `- 1` is needed here
            Some(max_clean_slot - 1)
        };
        self.rc.accounts.accounts_db.clean_accounts(max_clean_slot);
    }
}
```

Reply:
@ryoqun agreed, fixed!